// @flosportsinc/ng-fullscreen/esm5/src/common/ng-fullscreen.tokens.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export var FS_FULLSCREEN_REQUEST_EVENTS = new InjectionToken('fs.fullscreen.request');
/** @type {?} */
export var FS_FULLSCREEN_EXIT_EVENTS = new InjectionToken('fs.fullscreen.exit');
/** @type {?} */
export var FS_FULLSCREEN_CHANGE_EVENTS = new InjectionToken('fs.fullscreen.change');
/** @type {?} */
export var FS_FULLSCREEN_ELEMENT_ERROR_EVENTS = new InjectionToken('fs.fullscreen.error');
/** @type {?} */
export var FS_FULLSCREEN_ELEMENT = new InjectionToken('fs.fullscreen.element');
/** @type {?} */
export var FS_FULLSCREEN_ENABLED = new InjectionToken('fs.fullscreen.enabled');
/** @type {?} */
export var FS_FULLSCREEN_ENABLED_FUNC = new InjectionToken('fs.fullscreen.enabled-func');
/** @type {?} */
export var FS_FULLSCREEN_IOS_POLL_ENABLED = new InjectionToken('fs.fullscreen.ios.poll.enabled');
/** @type {?} */
export var FS_FULLSCREEN_IOS_POLL_MS = new InjectionToken('fs.fullscreen.ios.poll.ms');
