// @flosportsinc/ng-fullscreen/esm2015/src/common/ng-fullscreen.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FloFullscreenService } from './ng-fullscreen.service';
import { CommonModule } from '@angular/common';
import { FS_FULLSCREEN_REQUEST_EVENTS, FS_FULLSCREEN_EXIT_EVENTS, FS_FULLSCREEN_CHANGE_EVENTS, FS_FULLSCREEN_ELEMENT_ERROR_EVENTS, FS_FULLSCREEN_ELEMENT, FS_FULLSCREEN_ENABLED, FS_FULLSCREEN_ENABLED_FUNC, FS_FULLSCREEN_IOS_POLL_MS, FS_FULLSCREEN_IOS_POLL_ENABLED } from './ng-fullscreen.tokens';
import { DEFAULT_FS_FULLSCREEN_REQUEST_EVENTS, DEFAULT_FS_FULLSCREEN_EXIT_EVENTS, DEFAULT_FS_FULLSCREEN_CHANGE_EVENTS, DEFAULT_FS_FULLSCREEN_ELEMENT_ERROR_EVENTS, DEFAULT_FS_FULLSCREEN_ELEMENT, DEFAULT_FS_FULLSCREEN_ENABLED, DEFAULT_FS_FULLSCREEN_ENABLED_FUNC, DEFAULT_FS_FULLSCREEN_IOS_POLL_MS, DEFAULT_FS_FULLSCREEN_IOS_POLL_ENABLED } from './ng-fullscreen.tokens.defaults';
/**
 * @record
 */
export function FloFullscreenCommonModuleConfig() { }
if (false) {
    /** @type {?} */
    FloFullscreenCommonModuleConfig.prototype.ios;
}
/**
 * @record
 */
export function FloFullscreenCommonModuleIosPollingConfig() { }
if (false) {
    /** @type {?} */
    FloFullscreenCommonModuleIosPollingConfig.prototype.enabled;
    /** @type {?} */
    FloFullscreenCommonModuleIosPollingConfig.prototype.pollDurationMs;
}
const ɵ0 = DEFAULT_FS_FULLSCREEN_REQUEST_EVENTS, ɵ1 = DEFAULT_FS_FULLSCREEN_EXIT_EVENTS, ɵ2 = DEFAULT_FS_FULLSCREEN_CHANGE_EVENTS, ɵ3 = DEFAULT_FS_FULLSCREEN_ELEMENT_ERROR_EVENTS, ɵ4 = DEFAULT_FS_FULLSCREEN_ELEMENT, ɵ5 = DEFAULT_FS_FULLSCREEN_ENABLED, ɵ6 = DEFAULT_FS_FULLSCREEN_ENABLED_FUNC, ɵ7 = DEFAULT_FS_FULLSCREEN_IOS_POLL_MS, ɵ8 = DEFAULT_FS_FULLSCREEN_IOS_POLL_ENABLED;
export class FloFullscreenCommonModule {
    /**
     * @param {?} config
     * @return {?}
     */
    static config(config) {
        return {
            ngModule: FloFullscreenCommonModule,
            providers: [
                {
                    provide: FS_FULLSCREEN_IOS_POLL_ENABLED,
                    useValue: config.ios && typeof config.ios.enabled === 'boolean' ? config.ios.enabled : DEFAULT_FS_FULLSCREEN_IOS_POLL_ENABLED
                },
                {
                    provide: FS_FULLSCREEN_IOS_POLL_MS,
                    useValue: config.ios && typeof config.ios.pollDurationMs === 'number'
                        ? config.ios.pollDurationMs
                        : DEFAULT_FS_FULLSCREEN_IOS_POLL_MS
                }
            ]
        };
    }
}
FloFullscreenCommonModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                exports: [CommonModule],
                providers: [
                    FloFullscreenService,
                    { provide: FS_FULLSCREEN_REQUEST_EVENTS, useValue: ɵ0 },
                    { provide: FS_FULLSCREEN_EXIT_EVENTS, useValue: ɵ1 },
                    { provide: FS_FULLSCREEN_CHANGE_EVENTS, useValue: ɵ2 },
                    { provide: FS_FULLSCREEN_ELEMENT_ERROR_EVENTS, useValue: ɵ3 },
                    { provide: FS_FULLSCREEN_ELEMENT, useValue: ɵ4 },
                    { provide: FS_FULLSCREEN_ENABLED, useValue: ɵ5 },
                    { provide: FS_FULLSCREEN_ENABLED_FUNC, useFactory: ɵ6 },
                    { provide: FS_FULLSCREEN_IOS_POLL_MS, useValue: ɵ7 },
                    { provide: FS_FULLSCREEN_IOS_POLL_ENABLED, useValue: ɵ8 }
                ]
            },] }
];
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8 };
