// @acaisoft/angular-azure-maps/esm5/src/azure-map/utils/azure-map-lazy-loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
function Source() { }
/** @type {?} */
Source.prototype.type;
/** @type {?} */
Source.prototype.src;
/** @type {?} */
var sources = [
    {
        type: 'style',
        src: 'https://atlas.microsoft.com/sdk/css/atlas.min.css?api-version=1'
    },
    {
        type: 'script',
        src: 'https://atlas.microsoft.com/sdk/js/atlas.min.js?api-version=1'
    }
];
/**
 * @return {?}
 */
export function _window() {
    return window;
}
/** @typedef {?} */
var AssetType;
export { AssetType };
/**
 * @param {?} source
 * @param {?} type
 * @return {?}
 */
export function loadSingleAsset(source, type) {
    return new Promise(function (resolve, reject) {
        /** @type {?} */
        var createdElement;
        switch (type) {
            case 'script':
                createdElement = document.createElement('script');
                createdElement.src = source;
                break;
            case 'style':
                createdElement = document.createElement('link');
                createdElement.rel = 'stylesheet';
                createdElement.href = source;
                createdElement.type = 'text/css';
                break;
            default:
                reject('Wrong Type');
                break;
        }
        if (typeof (_window().atlas) !== 'undefined'
            && typeof (_window().atlas.Map) !== 'undefined') {
            resolve();
        }
        else {
            createdElement.onerror = reject;
            createdElement.onload = resolve;
            document.head.appendChild(createdElement);
        }
    });
}
/**
 * @return {?}
 */
export function azureMapLazyLoader() {
    return Promise.all(sources.map(function (source) { return loadSingleAsset(source.src, source.type); }));
}
