// @acaisoft/angular-azure-maps/esm5/src/azure-map/utils/load-map.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { azureMapLazyLoader } from './azure-map-lazy-loader';
import { BehaviorSubject, of, Subject } from 'rxjs';
import { fromPromise } from 'rxjs/internal-compatibility';
var LoadMapService = /** @class */ (function () {
    function LoadMapService() {
        this.isLoaded = false;
        this.isComponentLoaded = false;
        this.loadedSubject = new Subject();
        this.loadedComponenet = new BehaviorSubject(false);
    }
    /**
     * @return {?}
     */
    LoadMapService.prototype.load = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.isLoaded) {
            return fromPromise(azureMapLazyLoader().then(function () {
                _this.isLoaded = true;
                return _this.isLoaded;
            }));
        }
        else {
            return of(true);
        }
    };
    /**
     * @return {?}
     */
    LoadMapService.prototype.observableComponent = /**
     * @return {?}
     */
    function () {
        return this.loadedSubject.asObservable();
    };
    LoadMapService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    LoadMapService.ctorParameters = function () { return []; };
    return LoadMapService;
}());
export { LoadMapService };
if (false) {
    /** @type {?} */
    LoadMapService.prototype.isLoaded;
    /** @type {?} */
    LoadMapService.prototype.isComponentLoaded;
    /** @type {?} */
    LoadMapService.prototype.loadedSubject;
    /** @type {?} */
    LoadMapService.prototype.loadedComponenet;
}
