// @acaisoft/angular-azure-maps/esm2015/src/azure-map/utils/load-map.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { azureMapLazyLoader } from './azure-map-lazy-loader';
import { BehaviorSubject, of, Subject } from 'rxjs';
import { fromPromise } from 'rxjs/internal-compatibility';
export class LoadMapService {
    constructor() {
        this.isLoaded = false;
        this.isComponentLoaded = false;
        this.loadedSubject = new Subject();
        this.loadedComponenet = new BehaviorSubject(false);
    }
    /**
     * @return {?}
     */
    load() {
        if (!this.isLoaded) {
            return fromPromise(azureMapLazyLoader().then(() => {
                this.isLoaded = true;
                return this.isLoaded;
            }));
        }
        else {
            return of(true);
        }
    }
    /**
     * @return {?}
     */
    observableComponent() {
        return this.loadedSubject.asObservable();
    }
}
LoadMapService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
LoadMapService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    LoadMapService.prototype.isLoaded;
    /** @type {?} */
    LoadMapService.prototype.isComponentLoaded;
    /** @type {?} */
    LoadMapService.prototype.loadedSubject;
    /** @type {?} */
    LoadMapService.prototype.loadedComponenet;
}
