// @flosportsinc/ng-fullscreen/esm5/src/toggle/ng-fullscreen.toggle.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input, Inject, ChangeDetectorRef } from '@angular/core';
import { FloFullscreenService } from '../common/ng-fullscreen.service';
import { take } from 'rxjs/operators';
import { DOCUMENT } from '@angular/common';
// tslint:disable: readonly-keyword
// tslint:disable: no-object-mutation
var FloClickToEnterFullscreenDirective = /** @class */ (function () {
    function FloClickToEnterFullscreenDirective(fs, doc, cd) {
        this.fs = fs;
        this.doc = doc;
        this.cd = cd;
    }
    Object.defineProperty(FloClickToEnterFullscreenDirective.prototype, "floClickToEnterFullscreen", {
        get: /**
         * @return {?}
         */
        function () {
            return this._element;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val instanceof HTMLElement) {
                this._element = val;
            }
            else {
                this._element = this.doc.body;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FloClickToEnterFullscreenDirective.prototype.click = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.cd.detectChanges();
        this.fs.isNotFullscreen.pipe(take(1)).subscribe((/**
         * @param {?} _
         * @return {?}
         */
        function (_) {
            // should check for nested video elements to make iOS dev easier
            _this.fs.goFullscreen(_this.fs.extractVideoForIphoneIfRequired(_this.floClickToEnterFullscreen));
        }));
    };
    FloClickToEnterFullscreenDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[floClickToEnterFullscreen]',
                },] }
    ];
    /** @nocollapse */
    FloClickToEnterFullscreenDirective.ctorParameters = function () { return [
        { type: FloFullscreenService },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: ChangeDetectorRef }
    ]; };
    FloClickToEnterFullscreenDirective.propDecorators = {
        floClickToEnterFullscreen: [{ type: Input }],
        click: [{ type: HostListener, args: ['click', [],] }]
    };
    return FloClickToEnterFullscreenDirective;
}());
export { FloClickToEnterFullscreenDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FloClickToEnterFullscreenDirective.prototype._element;
    /**
     * @type {?}
     * @private
     */
    FloClickToEnterFullscreenDirective.prototype.fs;
    /**
     * @type {?}
     * @private
     */
    FloClickToEnterFullscreenDirective.prototype.doc;
    /**
     * @type {?}
     * @private
     */
    FloClickToEnterFullscreenDirective.prototype.cd;
}
var FloClickToExitFullscreenDirective = /** @class */ (function () {
    function FloClickToExitFullscreenDirective(fs) {
        this.fs = fs;
    }
    /**
     * @return {?}
     */
    FloClickToExitFullscreenDirective.prototype.click = /**
     * @return {?}
     */
    function () {
        this.fs.exitFullscreen();
    };
    FloClickToExitFullscreenDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[floClickToExitFullscreen]',
                },] }
    ];
    /** @nocollapse */
    FloClickToExitFullscreenDirective.ctorParameters = function () { return [
        { type: FloFullscreenService }
    ]; };
    FloClickToExitFullscreenDirective.propDecorators = {
        click: [{ type: HostListener, args: ['click', [],] }]
    };
    return FloClickToExitFullscreenDirective;
}());
export { FloClickToExitFullscreenDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FloClickToExitFullscreenDirective.prototype.fs;
}
