// @acaisoft/angular-azure-maps/esm5/src/azure-map/am.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AtlasMapComponent } from './atlas-map/atlas-map.component';
import { AtlasPopupDirective } from './directives/atlas-popup.directive';
import { LoadMapService } from './utils/load-map.service';
var AmModule = /** @class */ (function () {
    function AmModule() {
    }
    /**
     * @return {?}
     */
    AmModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: AmModule,
            providers: [
                LoadMapService
            ]
        };
    };
    AmModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                    ],
                    declarations: [AtlasMapComponent, AtlasPopupDirective],
                    exports: [
                        AtlasMapComponent,
                        AtlasPopupDirective,
                    ],
                    bootstrap: [AtlasMapComponent]
                },] },
    ];
    return AmModule;
}());
export { AmModule };
