// @flosportsinc/ng-fullscreen/esm2015/src/toggle/ng-fullscreen.toggle.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input, Inject, ChangeDetectorRef } from '@angular/core';
import { FloFullscreenService } from '../common/ng-fullscreen.service';
import { take } from 'rxjs/operators';
import { DOCUMENT } from '@angular/common';
// tslint:disable: readonly-keyword
// tslint:disable: no-object-mutation
export class FloClickToEnterFullscreenDirective {
    /**
     * @param {?} fs
     * @param {?} doc
     * @param {?} cd
     */
    constructor(fs, doc, cd) {
        this.fs = fs;
        this.doc = doc;
        this.cd = cd;
    }
    /**
     * @return {?}
     */
    get floClickToEnterFullscreen() {
        return this._element;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set floClickToEnterFullscreen(val) {
        if (val instanceof HTMLElement) {
            this._element = val;
        }
        else {
            this._element = this.doc.body;
        }
    }
    /**
     * @return {?}
     */
    click() {
        this.cd.detectChanges();
        this.fs.isNotFullscreen.pipe(take(1)).subscribe((/**
         * @param {?} _
         * @return {?}
         */
        _ => {
            // should check for nested video elements to make iOS dev easier
            this.fs.goFullscreen(this.fs.extractVideoForIphoneIfRequired(this.floClickToEnterFullscreen));
        }));
    }
}
FloClickToEnterFullscreenDirective.decorators = [
    { type: Directive, args: [{
                selector: '[floClickToEnterFullscreen]',
            },] }
];
/** @nocollapse */
FloClickToEnterFullscreenDirective.ctorParameters = () => [
    { type: FloFullscreenService },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: ChangeDetectorRef }
];
FloClickToEnterFullscreenDirective.propDecorators = {
    floClickToEnterFullscreen: [{ type: Input }],
    click: [{ type: HostListener, args: ['click', [],] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    FloClickToEnterFullscreenDirective.prototype._element;
    /**
     * @type {?}
     * @private
     */
    FloClickToEnterFullscreenDirective.prototype.fs;
    /**
     * @type {?}
     * @private
     */
    FloClickToEnterFullscreenDirective.prototype.doc;
    /**
     * @type {?}
     * @private
     */
    FloClickToEnterFullscreenDirective.prototype.cd;
}
export class FloClickToExitFullscreenDirective {
    /**
     * @param {?} fs
     */
    constructor(fs) {
        this.fs = fs;
    }
    /**
     * @return {?}
     */
    click() {
        this.fs.exitFullscreen();
    }
}
FloClickToExitFullscreenDirective.decorators = [
    { type: Directive, args: [{
                selector: '[floClickToExitFullscreen]',
            },] }
];
/** @nocollapse */
FloClickToExitFullscreenDirective.ctorParameters = () => [
    { type: FloFullscreenService }
];
FloClickToExitFullscreenDirective.propDecorators = {
    click: [{ type: HostListener, args: ['click', [],] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    FloClickToExitFullscreenDirective.prototype.fs;
}
